mitodl/mitxonline
# Migrate instructor data to individual selectable pages

# What are the relevant tickets?

Closes mitodl/hq#1543
#1773 

# Description (What does it do?)

Converts the current method of storing resource page instructor data (a stream block within the page marked Faculty Members) to a new method that uses individual pages for the instructors that can be linked to the resource page instead. Also contains some helper stuff to convert the data, so the existing data is not lost. 

#### Instructor pages

These are new and are designed to be located underneath an Instructor index page in Wagtail. The `configure_wagtail` command has been updated to create this index page if it doesn't exist.

Otherwise, instructor pages are normal Wagtail CMS pages. Some notes on fields:
* Title and Name are separate - in the case of duplicates, titles can be different to allow instructor pages to be more easily found in the CMS while the instructor's name remains the same. (This is especially important for the `convert_faculty_to_pages` command - see below.) 
* Name is what's used in the card on the page.
* There is a new Title field, this is displayed as a `<h4>` if present in the card.
* The short bio is analogous to the old description field.
* The long bio is not used currently.

#### `convert_faculty_to_pages`

This is a helper command that will load the data that's in the Faculty Members field in existing course pages, creates new Instructor Page instances for them, and then links them to the course page. It does _not_ operate on Program pages as we're not using them now. 

An Instructor Page is created for the faculty member listings _regardless of whether or not the instructor already exists in the system_ so we preserve whatever distinct imagery and copy is there from that particular page. The output of the command should tell you the title of the page that was created for the particular course page so you can find it later. These can be coalesced into a single instructor page at your leisure later. 

As part of the conversion process, the Faculty Members section is cleared for the page it's processing so duplicate cards aren't shown.

If there isn't an index page, the command will make one. 

#### Template changes

The `product_page.html` template is now set up to display the `Meet your instructors` section if there's linked Instructor Pages or old-style Faculty Members. Instructor pages will be displayed first. 

Instructor Pages have an additional Title field that wasn't in the old Faculty Members section - this is displayed as an `h4` under the name if there is one. 

No other changes for the cards - they should look the same as they did before. 

# How can this be tested?

1. Create a course page with some faculty members in it. 
2. Run `convert_faculty_to_pages`. This should convert the faculty member listings to pages, and apply them back to the course page.
3. View the course page. It should not be visibly different.
4. View the course page in the CMS. You should see the linked Instructor Pages and Faculty Members should be empty.
5. Explore the CMS - you should see an Instructors section with the new pages underneath. They should be Drafts.
